src/weak_memory.rs

```diff
@@ -0,0 +1,271 @@
+    }
+
+    /// Reads from the last store in modification order
+    pub fn read_from_last_store<'tcx>(&self, range: AllocRange, global: &GlobalState) {
+        let store_buffer = self.get_store_buffer(range);
+        let store_elem = store_buffer.buffer.back();
+        if let Some(store_elem) = store_elem {
+            let (index, clocks) = global.current_thread_state();
+            store_elem.load_impl(index, &clocks);
+        }
+    }
+
+    pub fn buffered_read<'tcx>(
+        &self,
+        range: AllocRange,
+        global: &GlobalState,
+        is_seqcst: bool,
+        rng: &mut (impl rand::Rng + ?Sized),
+        validate: impl FnOnce() -> InterpResult<'tcx>,
+    ) -> InterpResult<'tcx, Option<ScalarMaybeUninit<Tag>>> {
+        // Having a live borrow to store_buffer while calling validate_atomic_load is fine
+        // because the race detector doesn't touch store_buffer
+        let store_buffer = self.get_store_buffer(range);
+
+        let store_elem = {
+            // The `clocks` we got here must be dropped before calling validate_atomic_load
+            // as the race detector will update it
+            let (.., clocks) = global.current_thread_state();
+            // Load from a valid entry in the store buffer
+            store_buffer.fetch_store(is_seqcst, &clocks, &mut *rng)
+        };
+
+        // Unlike in write_scalar_atomic, thread clock updates have to be done
+        // after we've picked a store element from the store buffer, as presented
+        // in ATOMIC LOAD rule of the paper. This is because fetch_store
+        // require access to ThreadClockSet.clock, which is updated by the race detector
+        validate()?;
+
+        let loaded = store_elem.map(|store_elem| {
+            let (index, clocks) = global.current_thread_state();
+            store_elem.load_impl(index, &clocks)
+        });
+        Ok(loaded)
+    }
+
+    pub fn buffered_write<'tcx>(
+        &mut self,
+        val: ScalarMaybeUninit<Tag>,
+        range: AllocRange,
+        global: &GlobalState,
+        is_seqcst: bool,
+    ) -> InterpResult<'tcx> {
+        let (index, clocks) = global.current_thread_state();
+
+        let mut store_buffer = self.get_store_buffer_mut(range);
+        store_buffer.store_impl(val, index, &clocks.clock, is_seqcst);
+        Ok(())
+    }
+}
+
+const STORE_BUFFER_LIMIT: usize = 128;
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct StoreBuffer {
+    // Stores to this location in modification order
+    buffer: VecDeque<StoreElement>,
+}
+
+impl Default for StoreBuffer {
+    fn default() -> Self {
+        let mut buffer = VecDeque::new();
+        buffer.reserve(STORE_BUFFER_LIMIT);
+        Self { buffer }
+    }
+}
+
+impl<'mir, 'tcx: 'mir> StoreBuffer {
+    /// Selects a valid store element in the buffer.
+    /// The buffer does not contain the value used to initialise the atomic object
+    /// so a fresh atomic object has an empty store buffer until an explicit store.
```

Comment:
Okay, but what does that *mean*? Does this have any consequence for program behavior?

Reply:
An explanation was added.

Reply:
I elaborated the comment here and also explained why the workaround is needed in the test. The consequence for program behaviour is again - some allowed weak behaviours are not observable.

Is there anyway to add a shim for `Atomic*::new()`?

Reply:
> Is there anyway to add a shim for Atomic*::new()?

No, and that would also not make much sense -- that operation is not doing anything special in the abstract machine.
As I said, on the language level, there only are atomic accesses, not atomic objects. That is also how most formal memory models that I know of work (e.g. promising semantics), and it is how LLVM is specified. The fact that this weak memory emulation works differently is a weakness of the emulation. If the weakness only means we do not explore all possible executions, that is fine; it becomes a problem when this start throwing errors on otherwise well-defined code.